feat: add `subset` parameter to `DataFrame.dropna` to select which columns to consider

tswast opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes internal issue 366248570
🦕

tswast committed:
feat: add `subset` parameter to `DataFrame.dropna` to select which columns to consider

tswast commented:
Looks like some real test failures. I'll look at these later today.

Genesis929 committed:
Merge branch 'main' into b366248570-dropna-subset

tswast committed:
fix dropna with subset=None

tswast committed:
Merge remote-tracking branch 'origin/main' into b366248570-dropna-subset

tswast commented:
> Looks like some real test failures. I'll look at these later today.

Looks like there was a bug with `subset=None`. I believe this is now fixed.

Genesis929 commented on third_party/bigframes_vendored/pandas/core/frame.py:
```diff
@@ -1662,6 +1664,15 @@ def dropna(
             <BLANKLINE>
             [3 rows x 3 columns]
 
+        Define in which columns to look for missing values.
```
From doctest.
@@ -1,5 +1,5 @@
    -       name        toy       born
    -1    Batman  Batmobile 1940-04-25
    -2  Catwoman   Bullwhip        NaT
    +       name        toy        born
    +1    Batman  Batmobile  1940-04-25
    +2  Catwoman   Bullwhip        <NA>
     <BLANKLINE>
     [2 rows x 3 columns]

tswast replied:
Thanks, fixed! The diff got a little big because I pulled in https://github.com/googleapis/python-bigquery-dataframes/pull/987. It should be easier to review again once that is merged.

tswast committed:
refactor: remove circular dependencies preventing local doctest runs

With this change I can once again run

```
pytest --doctest-modules third_party/bigframes_vendored/pandas/core/frame.py
```

Note: having multiple `version.py` files should be fine. release-please
will update all such files it finds.

tswast committed:
Merge branch 'tswast-circular-import' into b366248570-dropna-subset

tswast committed:
fix doctest

tswast committed:
Revert "Merge branch 'tswast-circular-import' into b366248570-dropna-subset"

This reverts commit 57e8335af83558323d4df213854efc1edcdb35f0, reversing
changes made to 197074abc53c5a85c66ec1940e54ea3f2c9c3677.

tswast committed:
Reapply "Merge branch 'tswast-circular-import' into b366248570-dropna-subset"

This reverts commit 0f18294513a2a6a806f5fff4ff38646d4029a884.

tswast committed:
Merge branch 'main' into b366248570-dropna-subset

Genesis929 commented on bigframes/dataframe.py:
```diff
@@ -2033,8 +2034,23 @@ def dropna(
+            if subset is None:
+                subset_ids = None
+            elif not utils.is_list_like(subset):
+                subset_ids = [self._block.label_to_col_id[subset]]
```
The result of this seems to be a list of tuples, and don't work.

tswast replied:
This is supposed to cover the single column string case. A loop would give one character per iteration. I'll take a look at the error.

Genesis929 replied:
In single column case the subset_ids seems looks like [(col_id,)], I assume one loop is needed for the tuple?

Genesis929 replied:
From the defination ```def label_to_col_id(self) -> typing.Mapping[Label, typing.Sequence[str]]```, I think self._block.label_to_col_id[subset] will be a sequence(tuple)?  So I think a loop like for ```id in self._block.label_to_col_id[label]``` should be included?

tswast replied:
Chatted offline. `self._block.label_to_col_id[subset]` is a tuple. I've corrected.

tswast committed:
loop over tuple result

tswast committed:
Merge remote-tracking branch 'origin/main' into b366248570-dropna-subset

tswast committed:
Merge remote-tracking branch 'origin/main' into b366248570-dropna-subset
